src/util/pelz_io.c

```diff
@@ -201,140 +202,404 @@ int write_to_pipe(char *msg)
-      else
+      token_count++;
+    }
+  }
+  *num_tokens = token_count;
+  char **ret_tokens = (char **) malloc(token_count * sizeof(char *));
+
+  if (!ret_tokens)
+  {
+    pelz_log(LOG_ERR, "Unable to allocate memory.");
+    free(msg);
+    return (1);
+  }
+  char *save = msg;
+  char *token = strtok(msg, " ");
+
+  ret_tokens[0] = (char *) malloc(strlen(token) * sizeof(char) + 1);
+  if (!ret_tokens[0])
+  {
+    pelz_log(LOG_ERR, "Unable to allocate memory.");
+    free(save);
+    return (1);
+  }
+  memcpy(ret_tokens[0], token, strlen(token) + 1);  //copy the '\0'
+  for (size_t i = 1; i < token_count; i++)
+  {
+    char *token = strtok(NULL, " ");
+
+    if (token == NULL)
+    {
+      pelz_log(LOG_ERR, "Unable to tokenize pipe message: %s", msg);
+      for (size_t j = 0; j < i; j++)
       {
-        pelz_log(LOG_ERR, "Pipe command invalid: %s", msg);
-        return 0;
+        free(ret_tokens[j]);
       }
-      switch (opt)
+      free(ret_tokens);
+      free(save);
+      return (1);
+    }
+    ret_tokens[i] = (char *) malloc(strlen(token) * sizeof(char) + 1);
+    if (!ret_tokens[i])
+    {
+      pelz_log(LOG_ERR, "Unable to allocate memory.");
+      for (size_t j = 0; j < i; j++)
       {
-      case 'c':
-        len = strcspn(msg, "\n");
-        path = new_charbuf(len - 11); //the number 11 is used because it the number of chars in "pelz -l -c "
-        memcpy(path.chars, &msg[11], (path.len));
-        free_charbuf(&path);
-        pelz_log(LOG_INFO, "Load cert call not added");
-        return 0;
-      case 'p':
-        len = strcspn(msg, "\n");
-        path = new_charbuf(len - 11); //the number 11 is used because it the number of chars in "pelz -l -p "
-        memcpy(path.chars, &msg[11], (path.len));
-        free_charbuf(&path);
-        pelz_log(LOG_INFO, "Load private call not added");
-        return 0;
-      default:
-        pelz_log(LOG_ERR, "Pipe command invalid: %s", msg);
+        free(ret_tokens[j]);
+      }
+      free(ret_tokens);
+      free(save);
+      return (1);
+    }
+    memcpy(ret_tokens[i], token, strlen(token) + 1);  //copy the '\0'
+  }
+  if (strtok(NULL, " ") != NULL)
+  {
+    pelz_log(LOG_ERR, "Unable to tokenize pipe message: %s", msg);
+    for (size_t i = 0; i < token_count; i++)
+    {
+      free(ret_tokens[i]);
+    }
+    free(ret_tokens);
+    free(save);
+    return (1);
+  }
+  free(save);
+  *tokens = ret_tokens;
+  return (0);
+}
+
+int parse_pipe_message(char **tokens, size_t num_tokens, char **response)
+{
+  int ret;
+  char *path_ext = NULL;
+  charbuf key_id;
+  uint8_t *nkl_data = NULL;
+  size_t nkl_data_len = 0;
+  char *authString = NULL;
+  size_t auth_string_len = 0;
+  const char *ownerAuthPasswd = "";
+  size_t oa_passwd_len = 0;
+  uint8_t *data = NULL;
+  size_t data_length = 0;
+  uint64_t handle;
+
+  pelz_log(LOG_DEBUG, "Token num: %d", num_tokens);
+  if (num_tokens < 2)
+  {
+    *response = (char *) calloc(21, sizeof(char));
```

Comment:
I think the better way to handle response is have a ```#define RESPONSE_SIZE 1024``` and pass response in as an array. We can then use ```snprintf()``` to format response strings while making sure they stay within the RESPONSE_SIZE limit.